File: LeetCodeExample.Test/LinkedList/_00092_ReverseLinkedListII.py
# Given the head of a singly linked list and two integers left and right where left <= right, reverse the nodes of the list from position left to position right, and return the reversed list.

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        if left == right:
            return head
        before, begin, end, afterEnd = None, None, None, None
        i = 1
        current = head
        while current:
            if i == left - 1:
                before = current
            elif i == left:
                begin = current
            elif i == right:
                end = current
            elif i == right + 1:
                afterEnd = current
            i += 1
            current = current.next
        
        # Detach end from the list so the reverse doesn't include unwanted nodes
        end.next = None
        
        # Reverse the inner list
        prev = None
        current = begin
        while current:
            temp = current.next
            current.next = prev
            prev = current
            current = temp
        
        if before:
            before.next = prev
        begin.next = afterEnd
        return end if left == 1 else head

File: LeetCodeExample.Test/LinkedList/_00092_ReverseLinkedListII.cs
using LeetCodeExample.Test.Common;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using static LeetCodeExample.Test._00002_AddTwoNumbers_Test;

namespace LeetCodeExample.Test;

/// <summary>
//Given the head of a singly linked list and two integers left and right where left <= right, reverse the nodes of the list from position left to position right, and return the reversed list.
/// </summary>
public class _00092_ReverseLinkedListII
{
    public ListNode ReverseBetween(ListNode head, int left, int right)
    {
        if (head.next == null || left == right)
            return head;
        ListNode current = head;
        int i = 1;
        ListNode beforeReverse = null;
        ListNode reverseStart = null;
        ListNode reverseEnd = null;
        ListNode afterReverse = null;
        while (current != null)
        {
            if (i == left - 1)
                beforeReverse = current;
            else if (i == left)
                reverseStart = current;
            else if (i == right)
                reverseEnd = current;
            else if (i == right + 1)
                afterReverse = current;
            current = current.next;
            i++;
        }

        //Console.WriteLine($"b{beforeReverse.val}");
        //Console.WriteLine($"rs{reverseStart.val}");
        //Console.WriteLine($"re{reverseEnd.val}");
        //Console.WriteLine($"ar{afterReverse.val}");

        current = reverseStart;
        ListNode prev = afterReverse;
        for (int j = 0; j < right - left + 1; j++)
        {
            ListNode next = current.next;
            //Console.WriteLine($"Setting {current.val} to {prev.val}");
            current.next = prev;
            prev = current;
            current = next;
        }

        if (beforeReverse != null)
            beforeReverse.next = reverseEnd;
        // Return the new head
        return left == 1 ? reverseEnd : head;
    }
}